File: ios/ClawChatKit/Sources/ClawChatKit/Protocol/ClawChatMessage.swift
import Foundation

/// Discriminated union of all ClawChat wire protocol messages.
/// Decodes the `type` field first, then the appropriate struct.
public enum ClawChatMessage: Sendable {
    // Connection
    case appPaired(AppPaired)
    case appPairError(AppPairError)
    case appConnected(AppConnected)

    // Messaging
    case messageInbound(MessageInbound)
    case messageOutbound(MessageOutbound)
    case messageStream(MessageStream)
    case messageReasoning(MessageReasoning)
    case toolEvent(ToolEvent)

    // Control
    case typing(Typing)
    case presence(Presence)
    case statusResponse(StatusResponse)

    // Pairing
    case pairCode(PairCode)
    case devicesListResponse(DevicesListResponse)

    // Errors
    case error(ErrorMessage)

    // Ping/Pong
    case ping(PingMessage)
    case pong(PongMessage)

    // Unknown
    case unknown(type: String, raw: Data)
}

// MARK: - Ping / Pong

public struct PingMessage: BaseMessage, Sendable {
    public let type: String = "ping"
    public let id: String
    public let ts: Int64

    public init() {
        self.id = UUID().uuidString
        self.ts = Int64(Date().timeIntervalSince1970 * 1000)
    }
}

public struct PongMessage: BaseMessage, Sendable {
    public let type: String
    public let id: String
    public let ts: Int64
}

// MARK: - Decoding

private struct MessageEnvelope: Decodable {
    let type: String
}

extension ClawChatMessage {
    /// Decode a wire protocol message from raw JSON data.
    public static func decode(from data: Data) throws -> ClawChatMessage {
        let decoder = JSONDecoder()
        let envelope = try decoder.decode(MessageEnvelope.self, from: data)

        switch envelope.type {
        // Connection
        case "app.paired":
            return .appPaired(try decoder.decode(AppPaired.self, from: data))
        case "app.pair.error":
            return .appPairError(try decoder.decode(AppPairError.self, from: data))
        case "app.connected":
            return .appConnected(try decoder.decode(AppConnected.self, from: data))

        // Messaging
        case "message.inbound":
            return .messageInbound(try decoder.decode(MessageInbound.self, from: data))
        case "message.outbound":
            return .messageOutbound(try decoder.decode(MessageOutbound.self, from: data))
        case "message.stream":
            return .messageStream(try decoder.decode(MessageStream.self, from: data))
        case "message.reasoning":
            return .messageReasoning(try decoder.decode(MessageReasoning.self, from: data))
        case "tool.event":
            return .toolEvent(try decoder.decode(ToolEvent.self, from: data))

        // Control
        case "typing":
            return .typing(try decoder.decode(Typing.self, from: data))
        case "presence":
            return .presence(try decoder.decode(Presence.self, from: data))
        case "status.response":
            return .statusResponse(try decoder.decode(StatusResponse.self, from: data))

        // Pairing
        case "pair.code":
            return .pairCode(try decoder.decode(PairCode.self, from: data))
        case "devices.list.response":
            return .devicesListResponse(try decoder.decode(DevicesListResponse.self, from: data))

        // Errors
        case "error":
            return .error(try decoder.decode(ErrorMessage.self, from: data))

        // Ping/Pong
        case "ping":
            return .ping(try decoder.decode(PingMessage.self, from: data))
        case "pong":
            return .pong(try decoder.decode(PongMessage.self, from: data))

        default:
            return .unknown(type: envelope.type, raw: data)
        }
    }
}

File: clawchat/android/clawchat-kit/src/main/kotlin/com/clawchat/kit/protocol/ClawChatMessage.kt
package com.clawchat.kit.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.util.UUID

// MARK: - Ping / Pong

@Serializable
data class PingMessage(
    override val type: String = "ping",
    override val id: String = UUID.randomUUID().toString(),
    override val ts: Long = System.currentTimeMillis(),
) : BaseMessage

@Serializable
data class PongMessage(
    override val type: String,
    override val id: String,
    override val ts: Long,
) : BaseMessage

// MARK: - Discriminated union

sealed class ClawChatMessage {
    // Connection
    data class AppPairedMsg(val value: AppPaired) : ClawChatMessage()
    data class AppPairErrorMsg(val value: AppPairError) : ClawChatMessage()
    data class AppConnectedMsg(val value: AppConnected) : ClawChatMessage()

    // Messaging
    data class MessageInboundMsg(val value: MessageInbound) : ClawChatMessage()
    data class MessageOutboundMsg(val value: MessageOutbound) : ClawChatMessage()
    data class MessageStreamMsg(val value: MessageStream) : ClawChatMessage()
    data class MessageReasoningMsg(val value: MessageReasoning) : ClawChatMessage()
    data class ToolEventMsg(val value: ToolEvent) : ClawChatMessage()

    // Control
    data class TypingMsg(val value: Typing) : ClawChatMessage()
    data class PresenceMsg(val value: Presence) : ClawChatMessage()
    data class StatusResponseMsg(val value: StatusResponse) : ClawChatMessage()

    // Pairing
    data class PairCodeMsg(val value: PairCode) : ClawChatMessage()
    data class DevicesListResponseMsg(val value: DevicesListResponse) : ClawChatMessage()

    // Errors
    data class ErrorMsg(val value: ErrorMessage) : ClawChatMessage()

    // Ping/Pong
    data class PingMsg(val value: PingMessage) : ClawChatMessage()
    data class PongMsg(val value: PongMessage) : ClawChatMessage()

    // Unknown
    data class Unknown(val type: String, val rawJson: String) : ClawChatMessage()

    companion object {
        private val json = Json {
            ignoreUnknownKeys = true
            isLenient = true
        }

        /** Decode a wire protocol message from raw JSON string. */
        fun decode(rawJson: String): ClawChatMessage {
            val element = json.parseToJsonElement(rawJson)
            val type = element.jsonObject["type"]?.jsonPrimitive?.content
                ?: return Unknown("", rawJson)

            return try {
                when (type) {
                    // Connection
                    "app.paired" -> AppPairedMsg(json.decodeFromString<AppPaired>(rawJson))
                    "app.pair.error" -> AppPairErrorMsg(json.decodeFromString<AppPairError>(rawJson))
                    "app.connected" -> AppConnectedMsg(json.decodeFromString<AppConnected>(rawJson))

                    // Messaging
                    "message.inbound" -> MessageInboundMsg(json.decodeFromString<MessageInbound>(rawJson))
                    "message.outbound" -> MessageOutboundMsg(json.decodeFromString<MessageOutbound>(rawJson))
                    "message.stream" -> MessageStreamMsg(json.decodeFromString<MessageStream>(rawJson))
                    "message.reasoning" -> MessageReasoningMsg(json.decodeFromString<MessageReasoning>(rawJson))
                    "tool.event" -> ToolEventMsg(json.decodeFromString<ToolEvent>(rawJson))

                    // Control
                    "typing" -> TypingMsg(json.decodeFromString<Typing>(rawJson))
                    "presence" -> PresenceMsg(json.decodeFromString<Presence>(rawJson))
                    "status.response" -> StatusResponseMsg(json.decodeFromString<StatusResponse>(rawJson))

                    // Pairing
                    "pair.code" -> PairCodeMsg(json.decodeFromString<PairCode>(rawJson))
                    "devices.list.response" -> DevicesListResponseMsg(json.decodeFromString<DevicesListResponse>(rawJson))

                    // Errors
                    "error" -> ErrorMsg(json.decodeFromString<ErrorMessage>(rawJson))

                    // Ping/Pong
                    "ping" -> PingMsg(json.decodeFromString<PingMessage>(rawJson))
                    "pong" -> PongMsg(json.decodeFromString<PongMessage>(rawJson))

                    else -> Unknown(type, rawJson)
                }
            } catch (e: Exception) {
                Unknown(type, rawJson)
            }
        }

        /** Encode a BaseMessage to JSON string. */
        inline fun <reified T : BaseMessage> encode(message: T): String {
            return json.encodeToString(message)
        }
    }
}
